### PR TITLE
Magic Set and Pragma Fixes

### LINKED ISSUE #413: Magic sets and --debug-report segfaults
DebugReport segfaults in MagicSet when components are used. In evaluation/components_generic this seems to happen when `program->getRelation(relName)` is called and `relName` is a generated name like "SocialNet.r".

semantic/agg_checks2, which tests with relations that don't exist, also segfaults when the debug report attempts to show the result of transforming the non-existent relations.

I can silence the error with a nullptr check at MagicSet.cpp:412 and MagicSet.cpp:637, but it does look like modifying the traversal method might be better.

I used this to find examples with errors:
`SOUFFLE_CONFS="--debug-report=/tmp/a.html" SOUFFLE_CATEGORY="Syntactic,Semantic" make check`

### PR DESCRIPTION
**Pull request by azreika**
Fixes issues related to the magic set transformation and pragma declaratives:

- Pragma extended to allow for an empty argument by not specifying it. For example, we can write `.pragma "provenance"` in a datalog file to specify the `--provenance` command-line option with no arguments.

- Parts of the adornment stage have been refactored to allow for an easy change in the SIPS used. 

- Two new much more effective SIPS added: one that chooses the next atom with the maximum number of bound arguments, and another which chooses the one that maximises #bound_args/#args. Defaulting to the former now, but no benchmarking performed. Later addition might involve specifying the SIPS to use through the magic-transform option as well. tic-tac-toe.dl test went from 30 minutes to 10 seconds(!) on a local computer when run in interpreter mode with the new SIPS compared to the original (though the adornment stage does take slightly longer).

- Adornment has been removed from the debug report for the time being, for three main reasons:
   -  Placing it in the debug report meant it had to be computed at each transformation (even after the MST was performed) despite being unnecessary except for just before the magic transformation was executed.  With large programs containing high arity relations and using a complex SIP-Strategy, adornment may take a fair bit of time, meaning this would take up a lot of unnecessary computation time. 
   - Adornment was being calculated even when the magic-transform option was not set, meaning this computation time was added unnecessarily even then. This *may* be the cause for the hanging debug report in issue #411 despite not using the magic-transform option.
   - Since adornment was being calculated at every transformation stage, segfaults were occurring when non-existent relations were being used that would be handled by early transformations. By removing adornment from the debug-report, the adornment analysis only occurs after these problems are passed and we have a final valid program. This seems to fix #413 without any changes needed to the actual adornment code.

### CODE REVIEW

**Comment by b-scholz** [comment]
Abdul - you run out of build time in travis. Please can you contact Martin. He may have a workaround for this situation.

**Comment by mmcgr** [comment]
magic_hmmer, and maybe magic_liberal_identifiers , seem very slow with -c or --auto-schedule.
magic_access-policy -c is also slow, but not enough to reach the 10 minute timeout.

**Comment by b-scholz** [comment]
We plan to disable the auto-schedule feature anyway. Perhaps we do it now.

**Comment by b-scholz** [comment]
Martin - could you help. It seems we have a time-limit issue. I will try on my notebook as well.

**Comment by mmcgr** [comment]
The slower tests seem to be slower on linux as well, judging by their order in results. magic_liberal_identifiers -c doesn't finish until after 15 other tests on gcc/linux. On master it finishes sequentially. It's not 10 minute timeout slow, but it could help me duplicate the problem on linux.

**Comment by mmcgr** [comment]
I tried several compilers, but I haven't managed to duplicate the problem locally. I extended the timeout on travis to 20 minutes and it died with a 137 error, which usually means memory was exhausted.

**Comment by mmcgr** [comment]
I still can't duplicate the problem locally, but commit 30a2fe0: Refactored code involving SIPS, is the one that first started showing the problem on travis.

**Comment by mmcgr** [comment]
I think you should add a commit to this PR to remove the magic_hmmer and magic_liberal_identifiers tests. It seems those tests aren't failing because of this PR, it's just that this PR pushes their current flakiness over into consistent failure. (See #437 )

File: src/DebugReport.cpp [hunk 1]
```diff
@@ -230,15 +230,24 @@ void DebugReporter::generateDebugReport(
     DebugReportSection topsortSCCGraphSection =
             getCodeSection(id + "-topsort-scc-graph", "SCC Topological Sort Order", topsortSCCGraph.str());
 
+    // TODO (azreika): adornment output temporarily disabled; should only be computed just before MST
```
Comment (mmcgr) [inline src/DebugReport.cpp]: Could you remove this code instead of commenting it out? When it's time to re-enable, it can be merged back in.

**Pull request merged**